mqtt.go

```diff
@@ -19,598 +24,613 @@ const (
+	conn.mutex.Lock()
+	defer conn.mutex.Unlock()
+	conn.lastPacketID += 1
+	if conn.lastPacketID == 0 {
+		conn.lastPacketID = 1
 	}
+	return conn.lastPacketID
+}
 
-	mc.stream = packet.NewStream(mc.conn, mc.conn)
-
-	if err := mc.connect(); err != nil {
-		mc.conn.Close()
-		return nil, err
+// Returns the appropriate response channel for the channel type. For pubacks,
+// We will either create it or return the one that we stored previous
+func (conn *mqttConn) getResponseChan(p packet.Packet) mqttPacketChan {
+	//logInfo("[MQTT] Fetching response channel for %s\n", p.Type())
+	switch p.Type() {
+	case packet.CONNECT:
+		return conn.connRespCh
+	case packet.DISCONNECT:
+		// Disconnect is a special case. Server will close, so no response, but
+		// it's a packet type we expect, so don't let it fall to default
+		return nil
+	case packet.CONNACK:
+		return conn.connRespCh
+	case packet.SUBSCRIBE:
+		// Subscribes are blocking, but have packet ID's, and, theoretically,
+		// multiples can be sent, so we enable this at this level.
+		subPkt := p.(*packet.SubscribePacket)
+		subPkt.PacketID = conn.getPacketID()
+		returnCh := make(mqttPacketChan)
+		conn.packIDToChan[subPkt.PacketID] = returnCh
+		return returnCh
+	case packet.SUBACK:
+		// It's possible the SUBACK doesn't have a response channel if it
+		// was queued. In that case, we'll return a nil and handle it like
+		// PINGREQ or PUBACK
+		subAck := p.(*packet.SubackPacket)
+		return conn.packIDToChan[subAck.PacketID]
+	case packet.PUBLISH, packet.PINGREQ, packet.PUBACK, packet.PINGRESP:
+		// Since we have to respond differently to these, but we don't
+		// want to show an error
+		return nil
+	default:
+		//
+		logError("[MQTT] Unhandled packet type: %s", p.Type())
+		return nil
 	}
+}
 
-	subs := []mqttSubscription{
-		mqttSubscription{
-			topic:      fmt.Sprintf("/devices/%d/command", mc.dc.DeviceID),
-			msgHandler: mc.handleCommandMsg,
-		},
-		mqttSubscription{
-			topic:      fmt.Sprintf("/devices/%d/kv", mc.dc.DeviceID),
-			msgHandler: mc.handleKeyValueMsg,
-		},
-	}
+var (
+	infoLogger  = log.New(os.Stdout, "[MODE - INFO] ", log.LstdFlags)
+	errorLogger = log.New(os.Stderr, "[MODE - ERROR] ", log.LstdFlags)
+)
 
-	if err := mc.subscribe(subs); err != nil {
-		mc.conn.Close()
-		return nil, err
-	}
+func logInfo(format string, values ...interface{}) {
+	infoLogger.Printf(format+"\n", values...)
+}
 
-	mc.command = cmdQueue
-	mc.event = evtQueue
-	mc.bulkData = evtBulkDataQueue
-	mc.syncedBulkData = evtSyncedBulkDataCh
-	mc.kvSync = kvSyncQueue
-	mc.kvPush = kvPushQueue
-	mc.doPing = make(chan time.Duration, 1)
-	mc.stopPublisher = make(chan bool)
-	mc.err = make(chan error, 1)
-	mc.outPacket = make(chan packet.Packet, 1)
-	mc.puback = make(chan *packet.PubackPacket, 10)     // make sure this won't block
-	mc.pingresp = make(chan *packet.PingrespPacket, 10) // make sure this won't block
-
-	go mc.runPacketReader()
-	go mc.runPacketWriter()
-	go mc.runPublisher()
-	go mc.runPinger()
```

Comment:
Why did you remove pinger loop from `mqtt.go`?
PING is very specific to MQTT and all applications need to ping. It will be better if we can implement here.

Reply:
I had written it, but Ethan had suggested to move those things into the library in applications. But, everyone would have to implement it. So, I will move put it into some helper functions. Then, people can use it if they want, or write their own.